robonomics_liability/src/robonomics_liability/LiabilityExecutionThread.py

```diff
@@ -0,0 +1,85 @@
+
+        self.ipfs_client = ipfs_client
+        self.master_check_interval = master_check_interval
+        self.recording_topics = recording_topics
+
+        self.liability = liability
+
+        self.thread = threading.Thread(target=self.execute_liability, args=())
+        self.thread.daemon = True
+
+        self.liability_finish = False
+        self.__recorder = None
+        self.__liability_result_file = None
+        self.__liability_execution_namespace = "eth_{0}".format(self.liability.address)
+
+    def getLiabilityMsg(self):
+        return self.liability
+
+    def start(self):
+        self.thread.start()
+
+    def finish(self, success):
+        self.__recorder.stop()
+
+        ipfs_response = self.ipfs_client.add(self.__liability_result_file)
+        try:
+            self.liability.result = ipfs_response['Hash']
+        except TypeError:
+            rospy.logwarn('IPFS add proceeding error: %s', ipfs_response[1]['Message'])
+            self.liability.result = ipfs_response[0]['Hash']
+
+        self.liability_finish = True
+
+        result_msg = Result()
+        result_msg.liability = self.liability.address
+        result_msg.result = self.liability.result
+        result_msg.success = success
+        return result_msg
+
+    def __createRecorder(self, result_file):
+        recorder = Recorder(result_file,
+                            topics=self.recording_topics,
+                            namespace="/liability/{0}".format(self.__liability_execution_namespace),
+                            master_check_interval=self.master_check_interval)
+        return recorder
+
+    def execute_liability(self):
+        with TemporaryDirectory() as tmpdir:
+            rospy.logdebug('Temporary directory created: %s', tmpdir)
+            os.chdir(tmpdir)
+
+            rospy.logdebug('Getting objective %s...', self.liability.objective)
+            self.ipfs_client.get(self.liability.objective)
+            rospy.logdebug('Objective is written to %s', tmpdir + '/' + self.liability.objective)
+
+            self.__liability_result_file = os.path.join(tmpdir, 'result.bag')
+            rospy.logdebug('Start recording to %s...', self.__liability_result_file)
+            rospy.logdebug("Recording all topics: %s", (not self.recording_topics))
+
+            self.__recorder = self.__createRecorder(self.__liability_result_file)
+            self.__recorder.start()
+
+            rospy.logdebug('Rosbag recorder started')
+
+            objective_rosbag = get_rosbag_from_file(self.liability.objective)
+            if objective_rosbag is not None:
+                player = Player(objective_rosbag, self.__liability_execution_namespace)
+                player.start()
+                rospy.logdebug('Rosbag player started')
+            else:
+                rospy.logwarn('Skip playing objective using rosbag player in liability %s', self.liability.address)
+
+            while not self.liability_finish:
```

Comment:
Is it really needed? In my vision waiting here is legacy from previous version of code.
Only one problem in temporary directory creation, but you can create it once in constructor by https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp and remove after finish.